certgen/certgen.go

```diff
@@ -0,0 +1,178 @@
+	DNSPrefix string
+	// NotAfterInterval specifies an interval in which the NotAfter time of a
+	// certificate must fall.
+	//
+	// For example, if a certificate is being generated to be submitted to a
+	// temporal CT Log shard, then, in order to be accepted by the Log, its
+	// NotAfter value must fall within the Log's temporal range, so this field
+	// would be set to the temporal interval of the Log.  However, if a
+	// certificate is being generated to be submitted to a non-temporal CT Log,
+	// this field should be left unset/set to nil.
+	NotAfterInterval *interval.Interval
+}
+
+// CA is a Certificate Authority that issues certificates and certificate chains
+// using its SigningCert and SigningKey.
+type CA struct {
+	SigningCert *x509.Certificate
+	SigningKey  crypto.Signer
+	CertConfig  CertificateConfig
+}
+
+// IssueCertificate creates a new leaf certificate, issued by the key specified
+// in the SigningCert and SigningKey fields of the CA, and configured using the
+// CertConfig in the CA.
+func (ca *CA) IssueCertificate() (*x509.Certificate, error) {
+	key, err := rsa.GenerateKey(crand.Reader, keySizeBits)
+	if err != nil {
+		return nil, fmt.Errorf("error generating key pair: %s", err)
+	}
+
+	template, err := leafTemplate(ca.CertConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf template: %s", err)
+	}
+
+	leafDER, err := x509.CreateCertificate(crand.Reader, template, ca.SigningCert, key.Public(), ca.SigningKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf certificate: %s", err)
+	}
+
+	leaf, err := x509.ParseCertificate(leafDER)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing leaf certificate DER: %s", err)
+	}
+
+	return leaf, nil
+}
+
+func leafTemplate(c CertificateConfig) (*x509.Certificate, error) {
+	sn, err := randSerialNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	notAfter := randNotAfter(c.NotAfterInterval)
+
+	return &x509.Certificate{
+		SerialNumber: sn,
+		Subject: pkix.Name{
+			Country:            []string{c.SubjectCountry},
+			Organization:       []string{c.SubjectOrganization},
+			OrganizationalUnit: []string{c.SubjectOrganizationalUnit},
+			Locality:           []string{c.SubjectLocality},
+			CommonName:         c.SubjectCommonName,
+		},
+		NotBefore:          notAfter.Add(-certValidity),
+		NotAfter:           notAfter,
+		SignatureAlgorithm: c.SignatureAlgorithm,
+
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  false,
+		DNSNames:              []string{c.SubjectCommonName, extendedDNSSAN(c.DNSPrefix, c.SubjectCommonName)},
+	}, nil
+}
+
+func randSerialNumber() (*big.Int, error) {
+	i := big.NewInt(0)
+	return crand.Int(crand.Reader, i.SetUint64(math.MaxUint64))
+}
+
+func randNotAfter(notAfterInterval *interval.Interval) time.Time {
+	if notAfterInterval == nil {
+		// TODO(katjoyce): Make this random.
```

Comment:
Which bit? the value chosen for start (currently time.Now()) or the certValidity (which I think you want to keep fixed as 24hrs)